source/GStreamerWebAudioPlayerClient.cpp

```diff
@@ -0,0 +1,339 @@
+    std::string audioMimeType = gst_structure_get_name(structure);
+    audioMimeType = audioMimeType.substr(0, audioMimeType.find(' '));
+    std::string format = gst_structure_get_string(structure, "format");
+    firebolt::rialto::WebAudioPcmConfig pcm;
+    gint tmp;
+
+    if (format.empty())
+    {
+        GST_ERROR("Format not found in caps");
+        return result;
+    }
+
+    if (!gst_structure_get_int(structure, "rate", &tmp))
+    {
+        GST_ERROR("Rate not found in caps");
+        return result;
+    }
+    pcm.rate = tmp;
+
+    if (!gst_structure_get_int(structure, "channels", &tmp))
+    {
+        GST_ERROR("Rate not found in caps");
+        return result;
+    }
+    pcm.channels = tmp;
+
+    if (parseGstStructureFormat(format, pcm.sampleSize, pcm.isBigEndian, pcm.isSigned, pcm.isFloat))
+    {
+        GST_ERROR("Can't parse format or it is not supported");
+        return result;
+    }
+
+    mBackendQueue.callInEventLoop(
+        [&]()
+        {
+            firebolt::rialto::WebAudioConfig config{pcm};
+
+            uint32_t priority = 1;
+            if (mClientBackend->createWebAudioBackend(shared_from_this(), audioMimeType, priority, &config))
+            {
+                if (!mClientBackend->getDeviceInfo(m_preferredFrames, m_maximumFrames, m_supportDeferredPlay))
+                {
+                    GST_ERROR("GetDeviceInfo failed, could not process samples");
+                }
+                mIsOpen = true;
+            }
+            else
+            {
+                GST_ERROR("Could not create web audio backend");
+                mIsOpen = false;
+            }
+            result = mIsOpen;
+        });
+
+    return result;
+}
+
+bool GStreamerWebAudioPlayerClient::play()
+{
+    GST_DEBUG("entry:");
+
+    bool result = false;
+    mBackendQueue.callInEventLoop(
+        [&]()
+        {
+            if (mIsOpen)
+            {
+                result = mClientBackend->play();
+            }
+            else
+            {
+                GST_ERROR("No web audio backend");
+            }
+        });
+
+    return result;
+}
+
+bool GStreamerWebAudioPlayerClient::pause()
+{
+    GST_DEBUG("entry:");
+
+    bool result = false;
+    mBackendQueue.callInEventLoop(
+        [&]()
+        {
+            if (mIsOpen)
+            {
+                result = mClientBackend->pause();
+            }
+            else
+            {
+                GST_ERROR("No web audio backend");
+            }
+        });
+
+    return result;
+}
+
+bool GStreamerWebAudioPlayerClient::setEos()
+{
+    GST_DEBUG("entry:");
+
+    bool result = false;
+    mBackendQueue.callInEventLoop(
+        [&]()
+        {
+            if (mIsOpen)
+            {
+                if (mSampleDataBuffer.size())
+                {
+                    pushSamples();
+                }
+                m_pushSamplesTimer.cancel();
+                result = mClientBackend->setEos();
```

Comment:
What if we cant push all the samples here? We don't want to be setting eos until mSampleDataBuffer is empty.